From 377ec5c1ed050220e83735f4839d06a4dcdb95df Mon Sep 17 00:00:00 2001
From: Bugdebugger <bugdebugger@users.noreply.github.com>
Date: Sun, 5 Apr 2015 16:30:06 +0200
Subject: [PATCH 088/129] Fix exception on creation of directory hierachies

If XDG_CONFIG_HOME is set to a path which does not (fully) exist, pingus
fails to create the missing parent directories.

Fixes Pingus/pingus#133
---
 INSTALL.unix        | 16 ++++++++++------
 SConscript          |  4 ++++
 src/util/system.cpp | 13 +++++++------
 3 files changed, 21 insertions(+), 12 deletions(-)

diff --git a/INSTALL.unix b/INSTALL.unix
index 22db5eca7..cdc0ca7f0 100644
--- a/INSTALL.unix
+++ b/INSTALL.unix
@@ -5,14 +5,16 @@ Requirements:
 -------------
 To compile Pingus you need:
 
- g++           - http://gcc.gnu.org/
- SDL           - http://libsdl.org/
+ g++              - http://gcc.gnu.org/
+ SDL              - http://libsdl.org/
  SDL_mixer
  SDL_image
- boost         - http://www.boost.org
- boost-signals - http://www.boost.org
- libpng        - http://libpng.org/
- scons         - http://scons.org/
+ boost            - http://www.boost.org
+ boost-signals    - http://www.boost.org
+ boost-system     - http://www.boost.org
+ boost-filesystem - http://www.boost.org
+ libpng           - http://libpng.org/
+ scons            - http://scons.org/
 
 In most cases you will find all of these in your distribution and
 there shouln't be a need to compile anything manually.
@@ -26,6 +28,8 @@ In Ubuntu you can install everything by typing:
      libsdl-image1.2-dev \
      libboost-dev \
      libboost-signals-dev \
+     libboost-system-dev \
+     libboost-filesystem-dev \
      libpng12-dev \
      scons
 
diff --git a/SConscript b/SConscript
index 37558526d..adaedcb38 100644
--- a/SConscript
+++ b/SConscript
@@ -189,6 +189,10 @@ class Project:
     def configure_boost(self):
         if not self.conf.CheckLibWithHeader('boost_signals', 'boost/signals2.hpp', 'c++'):
             self.fatal_error += "  * library 'boost_signals2' not found\n"
+        if not self.conf.CheckLib('boost_system', '', '', 'c++'):
+            self.fatal_error += "  * library 'boost_system' not found\n"
+        if not self.conf.CheckLibWithHeader('boost_filesystem', 'boost/filesystem.hpp', 'c++'):
+            self.fatal_error += "  * library 'boost_filesystem' not found\n"
 
     def configure_png(self):
         if self.conf.CheckMyProgram('pkg-config'):
diff --git a/src/util/system.cpp b/src/util/system.cpp
index 7b6988c0e..3a5ef8221 100644
--- a/src/util/system.cpp
+++ b/src/util/system.cpp
@@ -36,6 +36,7 @@
 #  include <sys/types.h>
 #  include <unistd.h>
 #  include <errno.h>
+#  include <boost/filesystem.hpp>
 #else /* WIN32 */
 #  include <windows.h>
 #  include <direct.h>
@@ -246,11 +247,14 @@ void
 System::create_dir(std::string directory)
 {
 #ifndef WIN32
-  log_info("System::create_dir: %1%", directory);
-
   if (!exist(directory))
   {
-    if (mkdir(directory.c_str(), S_IRUSR | S_IWUSR | S_IXUSR | S_IRGRP | S_IXGRP) != 0)
+    std::string::iterator end = directory.end() - 1;
+    if(*end == '/') {
+      directory.erase(end);
+    }
+    log_info("System::create_dir: %1", directory);
+    if (!boost::filesystem::create_directories(directory.c_str()))
     {
       raise_exception(std::runtime_error, "System::create_dir: " << directory << ": " << strerror(errno));
     }
@@ -361,11 +365,8 @@ System::init_directories()
 
   std::string statdir  = get_userdir();
 
-  create_dir(statdir);
-
   // FIXME: We need a better seperation between user created levels,
   // FIXME: third party levels and levels from the base distri
-  create_dir(statdir + "levels/");
   create_dir(statdir + "levels/dist");
   create_dir(statdir + "themes/");
 
-- 
2.51.2

